# 1. Data fetching and cleaning
In this notebook we are going to fetch and perform an initial exploration of the dataset for the Git track of the Hércules Challenge. This dataset consists of 50 GitHub repositories.

## Setup
As always, we will begin the notebook by starting the logging system and importing some constants defined in the "\_\_init\_\_.py" file:

In [1]:
%run __init__.py

We will also define an auxiliary function to print the empty columns of the dataframes:

In [2]:
def print_empty_cols(df):
    for col in df.columns:
        print(col)
        print('-' * len(col))
        res = df[df[col] == ''].index
        print(f"{len(res)} articles have no value for column {col}")
        print(res)
        print('\n')


Finally, we will impor the bokeh library to show the charts in the notebook, and we will import the BokehHistogram class from the _herc\_common_ library to show our results:

In [3]:
from bokeh.io import output_notebook

output_notebook()

/home/alejandro/.envs/edma/lib/python3.7/site-packages/pandas/compat/__init__.py:117: UserWarning: Could not import the lzma module. Your installed Python is incomplete. Attempting to use lzma compression will result in a RuntimeError.
  warnings.warn(msg)


Loading BokehJS ...

In [4]:
from herc_common import BokehHistogram

hist = BokehHistogram(color_fill="mediumslateblue", color_hover="slateblue", bins=25)

## Getting the repository URLs

The URL for every repository of the Git dataset is stored in the '_data/repo\_urls.txt_' file. First of all, we will be loading all the urls from that file into a list:

In [5]:
REPO_URLS_FILE = 'repo_urls.txt'

with open(os.path.join(DATA_DIR, REPO_URLS_FILE), 'r') as f:
    repo_urls = [line.rstrip('\n') for line in f]

len(repo_urls)

50

In [6]:
repo_urls[0]

'https://github.com/cmungall/LIRICAL/'

## Parsing the data

Now that the URLs of every repository have been saved, we can start calling the [GitHub API](https://developer.github.com/v3/) to obtain information about each repo. Since the API has a limit of 50 requests per hour for non-authorized requests, and we will be making about 200 requests to fetch all the information, we will need to make use of a personal token to make calls to the API. More information about what a personal token is, and how to create one, can be accessed through the [following link](https://help.github.com/en/github/authenticating-to-github/creating-a-personal-access-token). In the next cell we will be asking for the token:

In [7]:
import getpass


try:
    from secret import GITHUB_TOKEN
except ModuleNotFoundError:
    GITHUB_TOKEN = getpass.getpass("Introduce your personal access token to acces the GitHub API: ")


We will now import a series of classes and functions that will be used to fetch information about a given repo and convert it to an instance of the _GitHubRepoData_ class. More information about these functions and classes can be accessed at the _src_ package:

In [8]:
from src import GitHubIssue, GitHubRepoData, parse_repo_url

Finally, we will be creating a list of GitHubRepoData instances with information about every repository from the dataset:

In [9]:
from tqdm import tqdm

git_dataset = []
pbar = tqdm(repo_urls)
for url in pbar:
    pbar.set_description(f"Processing repository: {url}")
    git_dataset.append(parse_repo_url(url, GITHUB_TOKEN))


Processing repository: https://github.com/mikel-egana-aranguren/FAIRifier:   8%|▊         | 4/50 [00:37<07:58, 10.39s/it]    

mikel-egana-aranguren-FAIRifier-d69d6a4/extensions/fairifier-rdf-extension/scripts/externals/angular.min.js


Processing repository: https://github.com/mikel-egana-aranguren/elda:  10%|█         | 5/50 [02:07<25:41, 34.25s/it]     

mikel-egana-aranguren-elda-e9bd9d3/elda-assets/src/main/webapp/openlayers/proj4js-combined.js
mikel-egana-aranguren-elda-e9bd9d3/elda-assets/src/main/webapp/velocity/js/vendor/proj4js-combined.js


Processing repository: https://github.com/pjotrp/genenetwork2:  34%|███▍      | 17/50 [05:38<06:46, 12.31s/it]                            

pjotrp-genenetwork2-86d885f/wqflask/wqflask/static/new/javascript/box.js


Processing repository: https://github.com/pauldevos/Basketball_Analytics: 100%|██████████| 50/50 [24:11<00:00, 29.03s/it]               


## Creating a dataframe

The instances created before provide a _to\_dict_ method that can be used to convert the class to a Python dict. This dict can be used to easily create a pandas DataFrame. This DataFrame will be used from now on to explore and interact with the dataset:

In [15]:
import pandas as pd

df = pd.DataFrame([repo.to_dict() for repo in git_dataset])
df.head()

gh_id                   name  \
0  216602979                LIRICAL   
1  199330464   wikidata_ontomatcher   
2  253207181          ro-crate-ruby   
3  212556220  Misc_Training_scripts   
4  155879756              FAIRifier   

                                         description             owner_name  \
0  LIkelihood Ratio Interpretation of Clinical Ab...               cmungall   
1          Matches ontology classes against wikidata               cmungall   
2  A Ruby gem for creating, manipulating and read...          markwilkinson   
3  A place for me to keep various miscellanelous ...          markwilkinson   
4                           A tool to make data FAIR  mikel-egana-aranguren   

                                           languages  \
0  {'Java': 492423, 'FreeMarker': 13149, 'Python'...   
1  {'Prolog': 14691, 'Makefile': 1472, 'Dockerfil...   
2                      {'Ruby': 52724, 'HTML': 1319}   
3                     {'Shell': 15815, 'Ruby': 9445}   
4  {'Java': 3514431, 'JavaScript': 967765, 'HTML'...   

                                         readme_text  \
0  LIRICAL\n\nLIkelihood Ratio Interpretation of ...   
1  Match an ontology to Wikidata\nThis applicatio...   
2  ro-crate-ruby\nThis is a WIP gem for creating,...   
3  Misc_Training_scripts\nA place for me to keep ...   
4  Dependencies:\n\nJava 8\nApache Ant\n\nBuildin...   

                                         issues_text  \
0                                                      
1  Will help with #1 and with https://github.com/...   
2                                                      
3                                                      
4                                                      

                                        commits_text  \
0  Merge pull request #442 from TheJacksonLaborat...   
1  Adding skos:altLabel\n\nhttps://github.com/cmu...   
2  Update LICENSE\nBump version\nTidy up and chec...   
3  added new cool 3-federated query\nfinished edi...   
4  Merge pull request #16 from Shamanou/developme...   

                                       comments_text  
0      note that the Jannovar dependency does not...  
1                                                     
2  *\n * Expands the tree to the target element a...  
3                                                     
4  *\n * This class contains all the code shared ...

## Data cleaning and feature engineering

First of all, we will be taking an initial look to the values from the dataset:

In [16]:
df.loc[:, df.columns != 'gh_id'].describe()

name                                        description  \
count          50                                                 43   
unique         50                                                 43   
top     FAIRifier  A Deep Learning Case Study to Generate Music S...   
freq            1                                                  1   

       owner_name                                    languages readme_text  \
count          50                                           50          50   
unique         27                                           50          49   
top      mbdebian  {'C#': 15985, 'CSS': 649, 'JavaScript': 34}               
freq            2                                            1           2   

       issues_text                                       commits_text  \
count           50                                                 50   
unique           6                                                 50   
top                 chore: a remark on how to use a non-empty tbox...   
freq            45                                                  1   

       comments_text  
count             50  
unique            30  
top                   
freq              21

As we can see above, altough all the repository names are unique, the other columns have some repeated values. Those repeated values could be empty or null values, so we are going to check if that is the case:

In [17]:
df[df.isnull().any(axis=1)]

gh_id                        name description   owner_name  \
11   57412597                hemodonacion        None    fanavarro   
23  161862375                   biohack18        None     leechuck   
24  171842501            biosample_jsonld        None      inutano   
29  151696606          JavaTermiteStarter        None  SciBiteLabs   
39   42526998        pythonontologysearch        None  jamesmalone   
40  238953196               my-react-form        None     twhetzel   
41  157134523  python-ecology-lesson-es-1        None    agbeltran   

                                            languages  \
11                        {'Perl': 97370, 'R': 36211}   
23                                  {'Groovy': 14573}   
24   {'Ruby': 9183, 'Shell': 3171, 'Dockerfile': 226}   
29                                    {'Java': 16959}   
39  {'JavaScript': 194124, 'Python': 5748, 'HTML':...   
40   {'JavaScript': 35896, 'CSS': 6855, 'HTML': 1271}   
41  {'Jupyter Notebook': 4579723, 'HTML': 60649, '...   

                                          readme_text issues_text  \
11  Lost in Translation\nStructure\nThis repositor...               
23                                          biohack18               
24  BioSample records in JSON-LD\nBioSample is a d...               
29  JavaTermiteStarter\nSet of basic code to get y...               
39                                                                  
40  React Form Demos\nDescription\nDemo project to...               
41  Canal de Slack en español\nLección de Data Car...               

                                         commits_text  \
11  delete files\nnew execution\nReadme updated\nr...   
23                     initial commit\nInitial commit   
24  mkdir\nremove postgres lib\nremove comment whi...   
29  Create README.md\nRefactoring\nRefactoring\nMe...   
39  remove test message\nChange call to ontology_i...   
40  remove file App.css\nadd more example forms\na...   
41  Update fechas-progreso.md\nMerge pull request ...   

                                        comments_text  
11                                                     
23                                                     
24                                                     
29  *\n * This class was created by simon on 03/09...  
39  ****\n****\n****\n****\n****\n****\n****\n****...  
40  \n      manifest.json provides metadata used w...  
41  \n  <div class="col-md-6">\n    <h3 id="syllab...

We can see from the output above that there are 7 repositories which do not have a description. We are going to replace those null values by an empty string:

In [18]:
df.fillna(value="", inplace=True)

Now, we are going to see know how many columns have an empty value:

In [19]:
print_empty_cols(df)

gh_id
-----
0 articles have no value for column gh_id
Int64Index([], dtype='int64')


name
----
0 articles have no value for column name
Int64Index([], dtype='int64')


description
-----------
7 articles have no value for column description
Int64Index([11, 23, 24, 29, 39, 40, 41], dtype='int64')


owner_name
----------
0 articles have no value for column owner_name
Int64Index([], dtype='int64')


languages
---------
0 articles have no value for column languages
Int64Index([], dtype='int64')


readme_text
-----------
2 articles have no value for column readme_text
Int64Index([8, 39], dtype='int64')


issues_text
-----------
45 articles have no value for column issues_text
Int64Index([ 0,  2,  3,  4,  5,  6,  8, 10, 11, 12, 14, 15, 16, 17, 18, 19, 20,
            21, 22, 23, 24, 25, 26, 27, 29, 30, 31, 32, 33, 34, 35, 36, 37, 38,
            39, 40, 41, 42, 43, 44, 45, 46, 47, 48, 49],
           dtype='int64')


commits_text
------------
0 articles have no value for column commits_text


/home/alejandro/.envs/edma/lib/python3.7/site-packages/pandas/core/ops/array_ops.py:253: FutureWarning: elementwise comparison failed; returning scalar instead, but in the future will perform elementwise comparison
  res_values = method(rvalues)


Most of the repositories (45 out of 50) don't have any issues, and 3 of them don't have a readme.

Finally, we are going to join both the description and the readme of each repository into a new column, and remove all extra spaces from that column:

In [20]:
import re

def clean(text):
    return re.sub('\s+', ' ', text).strip()

df['full_text'] = df["description"] + ". " + df["readme_text"]
            #". " + df["comments_text"] + \
            #". " + df["commits_text"]
empty_idx = df[df['full_text'] == '. '].index
df.loc[empty_idx, 'full_text'] = df.loc[empty_idx, 'commits_text']
df['full_text_cleaned'] = df['full_text'].apply(lambda x: clean(x))
df['full_text_cleaned'].loc[0][:500]

'LIkelihood Ratio Interpretation of Clinical AbnormaLities. LIRICAL LIkelihood Ratio Interpretation of Clinical AbnormaLities LIRICAL is designed to provide clincially interpretable computational analysis of phenotypic abnormalities (encoded using the Human Phenotype Ontology), optionally combined with an analysis of variants and genotypes if a VCF file is provided with the results of diagnostic gene panel, exome, or genome sequencing. Detailed documentation is available This is a useful website '

## Initial exploration

To finish with the contents of this notebook, we will make an initial exploration of the dataset.

### Text length

We are going to add a new column to the DataFrame with the length in number of characters of each repo's full text:

In [21]:
df['num_chars_text'] = df['full_text_cleaned'].apply(lambda x: len(x))
df['num_chars_text'].describe()

count       50.000000
mean      2490.040000
std       3423.445233
min         11.000000
25%        374.000000
50%       1770.500000
75%       3026.750000
max      20382.000000
Name: num_chars_text, dtype: float64

We can see that the average number of characters in the readme + description is about 2471, and the maximum length is 20382 characters. However, 75% of the repositories have a number of characters lower than 3027.

We are going to plot this distribution and save it to disk:

In [22]:
GIT_HIST_COLUMN = "num_chars_text"
GIT_HIST_TITLE = "Readme + Description length distribution"
GIT_HIST_XLABEL = "Readme and description length (# of characters)"
GIT_HIST_YLABEL = "Number of repositories"

hist.load_plot(df, GIT_HIST_COLUMN, GIT_HIST_TITLE,
          GIT_HIST_XLABEL, GIT_HIST_YLABEL, True)

In [23]:
hist.save_plot(os.path.join(NOTEBOOK_1_RESULTS_DIR, '1_Repo_text_length.svg'))

### Languages used

Finally, we are going to also explore the most used programming languages for each repository.

We will begin by creating an auxiliary function that will create an horizontal bar chart with the given data:

In [24]:
from bokeh.io import show
from bokeh.models import ColumnDataSource, HoverTool
from bokeh.palettes import Category20b_20
from bokeh.plotting import figure

def plot_horizontal_bar_chart(x_data, y_data, title, tooltip_x, tooltip_y,
                              sort=True, color_palette=Category20b_20):
    sorted_y_data = sorted(y_data, key=lambda x: x_data[y_data.index(x)]) if sort else y_data
    source = ColumnDataSource(data=dict(y_data=y_data, x_data=x_data, color=color_palette))
    p = figure(y_range=sorted_y_data, x_range=(0, max(x_data) * 1.1), plot_height=750, title=title,
               toolbar_location='right')
    p.hbar(y='y_data', right='x_data', height=0.7, color='color', legend_field="y_data",
           fill_alpha=0.75, hover_fill_alpha=1.0, source=source)
    p.ygrid.grid_line_color = None
    p.legend.orientation = "vertical"
    p.legend.location = "bottom_right"
    p.add_tools(HoverTool(tooltips=[(tooltip_y, "@y_data"), (tooltip_x, "@x_data")],
                          point_policy="follow_mouse"))

    show(p, notebook_handle=True)



Now, we can create new dataframe with each different programming language used in the dataset and the number of bytes belonging to that language for each repo:

In [25]:
languages_df = pd.DataFrame(df['languages'].values.tolist()).fillna(value=0, inplace=False)
languages_df.head()

Java  FreeMarker  Python   Prolog  Makefile  Dockerfile    Shell  \
0   492423.0     13149.0   849.0      0.0       0.0         0.0      0.0   
1        0.0         0.0     0.0  14691.0    1472.0       700.0    278.0   
2        0.0         0.0     0.0      0.0       0.0         0.0      0.0   
3        0.0         0.0     0.0      0.0       0.0         0.0  15815.0   
4  3514431.0         0.0     0.0      0.0       0.0         0.0  28836.0   

      Ruby      HTML  JavaScript  ...  Raku  TSQL  PowerShell  \
0      0.0       0.0         0.0  ...   0.0   0.0         0.0   
1      0.0       0.0         0.0  ...   0.0   0.0         0.0   
2  52724.0    1319.0         0.0  ...   0.0   0.0         0.0   
3   9445.0       0.0         0.0  ...   0.0   0.0         0.0   
4   2559.0  333450.0    967765.0  ...   0.0   0.0         0.0   

   Game Maker Language  Web Ontology Language  PLpgSQL  Gherkin  Common Lisp  \
0                  0.0                    0.0      0.0      0.0          0.0   
1                  0.0                    0.0      0.0      0.0          0.0   
2                  0.0                    0.0      0.0      0.0          0.0   
3                  0.0                    0.0      0.0      0.0          0.0   
4                  0.0                    0.0      0.0      0.0          0.0   

   ActionScript   C#  
0           0.0  0.0  
1           0.0  0.0  
2           0.0  0.0  
3           0.0  0.0  
4           0.0  0.0  

[5 rows x 43 columns]

By making use of the previously defined function and the new DataFrame we can plot the top 15 languages with the most number of bytes written in each repository:

In [26]:
from bokeh.palettes import Category20_15

NUM_LANGUAGES = 15

languages_sum = languages_df.sum().sort_values(ascending=False)
num_bytes = languages_df.sum()[:NUM_LANGUAGES]

plot_horizontal_bar_chart(list(languages_sum)[:NUM_LANGUAGES],
                          list(languages_sum.keys())[:NUM_LANGUAGES],
                          "Languages with the most number of bytes",
                          "Number of bytes", "Language",
                          color_palette=Category20_15)

We can see that most repositories have code written in Jupyter Notebooks, JavaScript and Java.

Although the number of bytes is an interesting measure, some languages tend to have a bigger repository size by nature. In the following cell we are going to select the most prominent language for each repository and plot the most used languages in the dataset:

In [27]:
from bokeh.palettes import Category20_10

most_used_languages = languages_df.idxmax(axis=1).value_counts()[:10]
plot_horizontal_bar_chart(list(most_used_languages),
                          list(most_used_languages.keys()),
                          "Top 10 most used languages",
                          "Number of repositories", "Language",
                          color_palette=Category20_10)

With this new measure the top order has changed a bit. Both JavaScript and Jupyter Notebooks remain in the top 5, but they have fallen some positions to both Java and Python.

## Saving the dataframe

Finally, we are going to serialize the dataframe so we can load it later on in the following notebooks:

In [28]:
GIT_DF_FILE_PATH = os.path.join(NOTEBOOK_1_RESULTS_DIR, 'git_dataframe.pkl')

df.to_pickle(GIT_DF_FILE_PATH)